techequipt/django-test
# feat: real estate listing

🚀 PR
-----
**This PR contains:**
- A django app including API that creates, updates, retrieves and deletes real estate properties with fields as in [requirement](https://github.com/techequipt/django-test#description)
- ModelViewSet for Real estate properties with appropriate filters, ordering and permissions (commit 2279b746f66fbd9a8d3ff19530bf774b72887219) setup
- Use `with_deleted=1` to include deleted properties in API results
- Some convenient seed data for quick check

Additional:
- `django-filter` is installed

_Needs for improvement:_ even when I can test API filtering and sorting completely fine on my browser, the unit tests for them seem not working, I noted it in my code (commit [78bf3e6](https://github.com/techequipt/django-test/commit/78bf3e6dce61a74ad2a52e9304b1f1d9f398416d#diff-8a1cf14b0073a13c44310989d4dcc250078d1bdcac1c4b8aa5fb6c06488591ecR44))

-----

How to test
-----
1. Boot up local environment using: `docker compose -f local.yml up`
2. Optionally add some seed data for quick check: `docker compose -f local.yml exec django python manage.py seed_real_estates` (or with a convenient container manager UI: `python manage.py seed_real_estates`)
3. Run `pytest` to make sure everything is fine: `docker compose -f local.yml exec django pytest` (or with a convenient container manager UI: `pytest`)
4. API endpoints are accessible at http://localhost:8000/api/real_estates/properties

Screenshots
-----
<details open><summary><b>Filter `for_sale`, `available` properties sorted by price desc.:</b></summary>
<img width="1053" alt="image" src="https://github.com/techequipt/django-test/assets/13363340/6da0f62d-4cd9-4367-b7f7-1889666cbe71">

<details><summary><b>Seed data:</b></summary>
<img width="369" alt="image" src="https://github.com/techequipt/django-test/assets/13363340/12b22fca-83a0-4c7e-a6ad-f632b62f27ca">
</details>

<details><summary><b>pytest:</b></summary>
<img width="353" alt="image" src="https://github.com/techequipt/django-test/assets/13363340/33094c6c-e4f6-462b-9424-fffe748ef1d9">
</details>
